File: Python/solutions/03-sum-nums.py
# Write a method that takes in an integer `num` and returns the sum of
# all integers between zero and num, up to and including `num`.
#
# Difficulty: easy.

# ITERATIVE SOLUTION
# def sum_nums(num):
#     result = 0
#     n = num
#     while n > 0:
#         result += n
#         n -= 1
#     return result

# RECURSIVE SOLUTION
def sum_nums(num):
    if num == 1: return num
    return num + sum_nums(num - 1)

# These are tests to check that your code is working. After writing
# your solution, they should all print True.

print('sum_nums(1) == 1: ' + str(sum_nums(1) == 1))
print('sum_nums(2) == 3: ' + str(sum_nums(2) == 3))
print('sum_nums(3) == 6: ' + str(sum_nums(3) == 6))
print('sum_nums(4) == 10: ' + str(sum_nums(4) == 10))
print('sum_nums(5) == 15: ' + str(sum_nums(5) == 15))

File: Ruby/solutions (from jayqui)/03-sum-nums.rb
# Write a method that takes in an integer `num` and returns the sum of
# all integers between zero and num, up to and including `num`.
#
# Difficulty: easy.

### A pretty transparent solution
# def sum_nums(num)
#   sum = 0
#   (0..num).each {|n| sum += n}
#   sum
# end

### The Ruby-tacular way. Under the hood it does exactly the same as the above.
def sum_nums(num)
  (0..num).inject(:+)
end

  ### Line 15 is shorthand for line 20:
  # def sum_nums(num)
  #   (0..num).inject(0) {|accum, n| accum + n}
  # end

# These are tests to check that your code is working. After writing
# your solution, they should all print true.

puts('sum_nums(1) == 1: ' + (sum_nums(1) == 1).to_s)
puts('sum_nums(2) == 3: ' + (sum_nums(2) == 3).to_s)
puts('sum_nums(3) == 6: ' + (sum_nums(3) == 6).to_s)
puts('sum_nums(4) == 10: ' + (sum_nums(4) == 10).to_s)
puts('sum_nums(5) == 15: ' + (sum_nums(5) == 15).to_s)
